src/sage/rings/function_field/drinfeld_modules/drinfeld_module.py

```diff
@@ -1089,19 +1400,137 @@ def j_invariant(self):
+            ...
+            TypeError: list or tuple parameter must contain tuples or lists
+            sage: phi.j_invariant(((1, 2, 3, 4, 5), (2, 1)))
+            Traceback (most recent call last):
+            ...
+            ValueError: components of tuple or list parameter have incorrect length
+            sage: phi.j_invariant(((1, 'x'), (2, 3, 8)))
+            Traceback (most recent call last):
+            ...
+            TypeError: components of tuple or list parameter must contain only integers
+            sage: phi.j_invariant(((1, 2), (2, 3, 'x')))
+            Traceback (most recent call last):
+            ...
+            TypeError: components of tuple or list parameter must contain only integers
+            sage: phi.j_invariant(((1, 2), (4, 3, 7)))
+            Traceback (most recent call last):
+            ...
+            ValueError: parameter does not satisfy the weight-0 condition
+            sage: phi.j_invariant(((1, 2), (4, 3, 7)), check=False)
+            1/T^13
         """
-        self._check_rank_two()
-        g = self.coefficient(1)
-        delta = self.coefficient(2)
+        r = self._gen.degree()
         q = self._Fq.order()
-        return (g**(q+1)) / delta
+        if parameter is None:
+            if r != 2:
+                raise TypeError("parameter must be different from "
+                                 "NoneType if the rank is greater than 2")
+            return self._gen[1]**(q+1)/self._gen[2]
+        if parameter in ZZ:
+            if parameter <= 0 or parameter >= r:
+                raise ValueError("integer parameter must be >= 1 and < the "
+                                 f"rank (={r})")
+            dk = Integer((q**r - 1)/(q**gcd(parameter, r) - 1))
+            dr = Integer((q**parameter - 1)/(q**gcd(parameter, r) - 1))
+            return self._gen[parameter]**dk / self._gen[-1]**dr
+        elif isinstance(parameter, (tuple, list)):
+            if len(parameter) != 2:
+                raise ValueError("list or tuple parameter must be of length 2")
+            if not isinstance(parameter[0], (tuple, list)) \
+                    or not isinstance(parameter[1], (tuple, list)):
+                raise TypeError("list or tuple parameter must contain tuples "
+                                "or lists")
+            if not len(parameter[0]) < r or\
+                       not len(parameter[1]) == len(parameter[0]) + 1:
+                raise ValueError("components of tuple or list parameter have "
+                                 "incorrect length")
+            if not all(p in ZZ for p in parameter[0])\
+                    or not all(p in ZZ for p in parameter[1]):
+                raise TypeError("components of tuple or list parameter must "
+                                "contain only integers")
+            # Check that the weight-0 condition is satisfied:
+            # d_1 (q - 1) + d_2 (q^2 - 1) + ... + d_{r-1} (q^{r-1} - 1)
+            #    = d_r (q^r - 1)
```

Comment:
Generally, what about indenting math equations and inequations? Here, I am using an indentation of two.

```suggestion
            # Check that the weight-0 condition is satisfied:
            #   d_1 (q - 1) + ... + d_{r-1} (q^{r-1} - 1)
            #   = d_r (q^r - 1)
```